### ISSUE TITLE
Memory leak with big-endian data

### ISSUE THREAD

**Issue by saimn** [pydata/bottleneck#351]
**Describe the bug**

Investigating a memory leak with a sigma-clipping function which is using bottleneck to compute a `nanmedian` and ` nanstd` (https://github.com/astropy/astropy/issues/10553).  The leak actually happens in bottleneck, when the data is big-endian (which is the case with FITS data, a widely used format in astronomy).

**To Reproduce**

Reproduced on linux/Python 3.8 with a file from the original but report:
```
import gc
import sys
import numpy as np
import psutil
import bottleneck
from astropy.io import fits
from astropy.stats import sigma_clip

data = fits.getdata("templatealigned.fits")
data = data[np.isfinite(data)]  # to avoid warnings with NaNs
if len(sys.argv) > 1:
    data = data.astype(sys.argv[1])

print(data.dtype)
p = psutil.Process()
mem = p.memory_info().rss

cenfunc = bottleneck.nanmedian
stdfunc = bottleneck.nanstd

for i in range(20):
    _ = sigma_clip(data, sigma=3.0, cenfunc=cenfunc, stdfunc=stdfunc)
    print(f'iter {i+1:02d} : {(p.memory_info().rss - mem) / 1024**2:.2f}')
    gc.collect()
```

By default data type is `>f4` and it leaks:
```
❯ python leakreproduction.py
>f4
iter 01 : 66.29
iter 02 : 127.34
iter 03 : 177.72
iter 04 : 208.50
iter 05 : 258.86
iter 06 : 289.51
iter 07 : 340.00
iter 08 : 370.79
iter 09 : 421.14
iter 10 : 451.93
iter 11 : 502.29
iter 12 : 532.81
iter 13 : 583.29
iter 14 : 613.95
iter 15 : 664.56
iter 16 : 695.09
iter 17 : 745.70
iter 18 : 776.23
iter 19 : 826.84
iter 20 : 857.50
```
Casting to little endian fixes the leak:
```
❯ python leakreproduction.py '<f4'       
float32
iter 01 : 28.34
iter 02 : 41.12
iter 03 : 41.12
iter 04 : 46.21
iter 05 : 46.21
iter 06 : 41.17
iter 07 : 41.17
iter 08 : 46.26
iter 09 : 46.26
iter 10 : 41.17
iter 11 : 41.17
iter 12 : 46.26
iter 13 : 46.26
iter 14 : 41.17
iter 15 : 41.17
iter 16 : 46.26
iter 17 : 46.26
iter 18 : 41.17
iter 19 : 41.17
iter 20 : 46.26
```
And I get the same kind of numbers with `<f8` vs `>f8`.

I guess when the "slow" function is used when data is big-endian, it is missing a `Py_DECREF(a);` here (trying this locally fixes the issue, but the good fix probably requires more) :

https://github.com/pydata/bottleneck/blob/7c685ae3ac4e60ed0a214b8a6ba23a800b10dd76/bottleneck/src/reduce_template.c#L1789-L1792

**Comment by qwhelan** [comment]
Apologies for the delay and thanks for reporting.

Is there an example `.fits` file that I can use? It seems `astropy` doesn't provide the example file, so I cannot reproduce.

**Comment by saimn** [comment]
Hi @qwhelan ,
Thanks for investigating this. It is reproducible directly with a Numpy array:
```
In [7]: import gc
   ...: import sys
   ...: import numpy as np
   ...: import psutil
   ...: import bottleneck
   ...: from astropy.stats import sigma_clip
   ...: 
   ...: data = np.random.randn(2048, 2048)
   ...: data = data.astype('>f4')
   ...: 
   ...: print(data.dtype)
   ...: p = psutil.Process()
   ...: mem = p.memory_info().rss
   ...: cenfunc = bottleneck.nanmedian
   ...: stdfunc = bottleneck.nanstd
   ...: for i in range(20):
   ...:     _ = sigma_clip(data, sigma=3.0, cenfunc=cenfunc, stdfunc=stdfunc)
   ...:     print(f'iter {i+1:02d} : {(p.memory_info().rss - mem) / 1024**2:.2f}')
   ...:     gc.collect()
   ...: 
>f4
iter 01 : 79.79
iter 02 : 159.52
iter 03 : 223.44
iter 04 : 287.21
iter 05 : 351.14
iter 06 : 414.91
iter 07 : 478.85
iter 08 : 542.61
```

**Comment by qwhelan** [comment]
Keeping this open until released (probably as 1.3.3)

**Comment by rdbisme** [comment]
This fix should have landed in v1.3.4
